### PR TITLE
fix: GoCardless payment creation exception when no mandate available

### PR DESCRIPTION
**Pull request by ancorcruz**
GoCardless payment creation does not raise error when no mandate available

Instead of raising an error when a mandate is not available at GoCardless to create a new payment,

We notify the client with a meaningful error message, flag the invoice as payment failed and return a failure result.

## Context

Given a customer does not have a GoCardless mandate available to process an invoice payment, `Invoices::Payments::GocardlessCreateJob` was raising an exception, moving the job to sidekiq dead jobs queue and notifying the client with not helpful information `"One of your parameters was incorrectly typed"`.

```
job_name: "Invoices::Payments::GocardlessCreateJob"
error_class: "GoCardlessPro::InvalidApiUsageError"
error_message: "One of your parameters was incorrectly typed"
```

## Description

This change fixes this issue exiting the job with a service failure result, notifying the client pointing the customer has no available mandate for the payment and updates the invoice as payment failed.

### CODE REVIEW

File: app/services/invoices/payments/gocardless_service.rb [hunk 1]
```diff
@@ -5,6 +5,18 @@ module Payments
     class GocardlessService < BaseService
       include Customers::PaymentProviderFinder
 
+      class MandateNotFoundError < StandardError
+        DEFAULT_MESSAGE = 'No mandate avaiable for payment'
+
+        attr_reader :code
+
+        def initialize(msg = DEFAULT_MESSAGE)
+          @code = '999'
```
Comment (ancorcruz) [inline app/services/invoices/payments/gocardless_service.rb:14]: no idea what error code use here...

**Comment by ancorcruz** [comment]
@michaelponrajah @sarkissianraffi What do you think about the error message?

File: app/services/invoices/payments/gocardless_service.rb [hunk 2]
```diff
@@ -5,6 +5,18 @@ module Payments
     class GocardlessService < BaseService
       include Customers::PaymentProviderFinder
 
+      class MandateNotFoundError < StandardError
+        DEFAULT_MESSAGE = 'No mandate avaiable for payment'
```
Comment (ivannovosad) [inline app/services/invoices/payments/gocardless_service.rb:9]:
```suggestion
        DEFAULT_MESSAGE = 'No mandate available for payment'
```

**Review by ivannovosad — CHANGES_REQUESTED** [review]

File: app/services/invoices/payments/gocardless_service.rb (hunk 2 above)
Comment (ancorcruz) [reply app/services/invoices/payments/gocardless_service.rb]: Thanks for catching the typo.

File: app/services/invoices/payments/gocardless_service.rb [hunk 3]
```diff
@@ -5,6 +5,18 @@ module Payments
     class GocardlessService < BaseService
       include Customers::PaymentProviderFinder
 
+      class MandateNotFoundError < StandardError
+        DEFAULT_MESSAGE = 'No mandate available for payment'
+
+        attr_reader :code
+
+        def initialize(msg = DEFAULT_MESSAGE)
+          @code = '999'
```
Comment (ivannovosad) [inline app/services/invoices/payments/gocardless_service.rb:14]: We can use something like `No mandate error` or `no_mandate_error`.

File: spec/services/invoices/payments/gocardless_service_spec.rb [hunk 4]
```diff
@@ -174,6 +174,44 @@
...
+          .and_raise(GoCardlessPro::Error.new('code' => 'code', 'message' => 'error'))
+      end
+
+      it 'delivers an error webhook' do
+        result = gocardless_service.create
+
+        aggregate_failures do
+          expect(result).not_to be_success
+          expect(result.error).to be_a(BaseService::ServiceFailure)
+          expect(result.error.code).to eq('999')
```
Comment (ivannovosad) [inline spec/services/invoices/payments/gocardless_service_spec.rb:196]: In case of changing the code we'll need to change it here as well.

**Review by ivannovosad — CHANGES_REQUESTED** [review]

File: app/services/invoices/payments/gocardless_service.rb (hunk 3 above)
Comment (ancorcruz) [reply app/services/invoices/payments/gocardless_service.rb:14]: looks good

**Review by julienbourdeau — APPROVED** [review]

**Pull request merged**